File: src/NodeParser/ImportTypeNodeParser.ts
import ts from "typescript";
import { Context, type NodeParser } from "../NodeParser.js";
import type { SubNodeParser } from "../SubNodeParser.js";
import type { BaseType } from "../Type/BaseType.js";

export class ImportTypeNodeParser implements SubNodeParser {
    public constructor(
        protected typeChecker: ts.TypeChecker,
        protected childNodeParser: NodeParser,
    ) {}

    public supportsNode(node: ts.ImportTypeNode): boolean {
        return node.kind === ts.SyntaxKind.ImportType;
    }

    public createType(node: ts.ImportTypeNode, context: Context): BaseType {
        // Resolve the symbol from the qualifier (e.g., `MyType` in `import("./module").MyType`)
        const symbolLocation = node.qualifier ?? node;
        const typeSymbol = this.typeChecker.getSymbolAtLocation(symbolLocation)!;

        // Handle transitive re-exports
        if (typeSymbol.flags & ts.SymbolFlags.Alias) {
            const aliasedSymbol = this.typeChecker.getAliasedSymbol(typeSymbol);
            return this.childNodeParser.createType(
                aliasedSymbol.declarations![0],
                this.createSubContext(node, context),
            );
        }

        return this.childNodeParser.createType(typeSymbol.declarations![0], this.createSubContext(node, context));
    }

    protected createSubContext(node: ts.ImportTypeNode, parentContext: Context): Context {
        const subContext = new Context(node);
        if (node.typeArguments?.length) {
            for (const typeArg of node.typeArguments) {
                subContext.pushArgument(this.childNodeParser.createType(typeArg, parentContext));
            }
        }
        return subContext;
    }
}

File: internal/parser/import_type_node_parser.go
package parser

import (
	"github.com/microsoft/typescript-go/shim/ast"
	"github.com/microsoft/typescript-go/shim/checker"

	"github.com/vega/ts-json-schema-generator/internal/types"
)

// ImportTypeNodeParser resolves `import("./module").MyType` type nodes
// (src/NodeParser/ImportTypeNodeParser.ts).
type ImportTypeNodeParser struct {
	typeChecker     *checker.Checker
	childNodeParser NodeParser
}

func NewImportTypeNodeParser(typeChecker *checker.Checker, childNodeParser NodeParser) *ImportTypeNodeParser {
	return &ImportTypeNodeParser{typeChecker: typeChecker, childNodeParser: childNodeParser}
}

func (p *ImportTypeNodeParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindImportType
}

func (p *ImportTypeNodeParser) CreateType(node *ast.Node, ctx *Context, _ *types.ReferenceType) types.Type {
	// Resolve the symbol from the qualifier (e.g., `MyType` in
	// `import("./module").MyType`).
	symbolLocation := node.AsImportTypeNode().Qualifier
	if symbolLocation == nil {
		symbolLocation = node
	}
	typeSymbol := p.typeChecker.GetSymbolAtLocation(symbolLocation)

	// Handle transitive re-exports.
	if typeSymbol.Flags&ast.SymbolFlagsAlias != 0 {
		aliasedSymbol := p.typeChecker.GetAliasedSymbol(typeSymbol)
		return p.childNodeParser.CreateType(aliasedSymbol.Declarations[0], p.createSubContext(node, ctx), nil)
	}

	return p.childNodeParser.CreateType(typeSymbol.Declarations[0], p.createSubContext(node, ctx), nil)
}

func (p *ImportTypeNodeParser) createSubContext(node *ast.Node, parentContext *Context) *Context {
	subContext := NewContext(node)
	for _, typeArg := range node.TypeArguments() {
		subContext.PushArgument(p.childNodeParser.CreateType(typeArg, parentContext, nil))
	}
	return subContext
}
